File: frontend/src/services/api/academicRequestService.ts
import apiClient from './authService';

// Types
export interface AcademicRequest {
    id: number;
    studentId: number;
    studentCode: string;
    studentName: string;
    studentEmail: string;
    studentAvatar?: string;
    studentMajor?: string;
    studentSpecialization?: string;
    studentSubSpecialization?: string;
    requestType: string;
    requestTypeLabel: string;
    requestTitle: string;
    semesterId?: number;
    semesterCode?: string;
    semesterName?: string;
    courseId?: number;
    courseCode?: string;
    courseName?: string;
    classSectionId?: string;
    className?: string;
    toClassName?: string;
    toMajor?: string;
    toSpecialization?: string;
    toSubSpecialization?: string;
    reason: string;
    note?: string;
    fileUrl?: string;
    status: string;
    statusLabel: string;
    startDate?: string;
    dueDate?: string;
    isWithinDeadline?: boolean;
    isTransferPossible?: boolean;
    transferError?: string;
    isApprovable?: boolean;
    validationMessage?: string;
    approverId?: number;
    approverName?: string;
    approvedAt?: string;
    approverNote?: string;
    approverAvatar?: string;
    createdAt: string;
    updatedAt: string;
}

export interface AcademicRequestType {
    value: string;
    label: string;
    description?: string;
    deadlineRule: string;
    startDate?: string;
    dueDate?: string;
    canSubmit: boolean;
    requiresClassSection?: boolean;
}

export interface PageResponse<T> {
    content: T[];
    totalPages: number;
    totalElements: number;
    size: number;
    number: number;
}

export interface CreateAcademicRequestPayload {
    requestType: string;
    requestTitle?: string;
    semesterId?: number;
    courseId?: number;
    classSectionId?: string;
    toClassName?: string;
    toMajor?: string;
    toSpecialization?: string;
    toSubSpecialization?: string;
    reason: string;
    note?: string;
}

export interface DeadlineCheckResponse {
    canSubmit: boolean;
    message: string;
    startDate?: string;
    dueDate?: string;
}

export interface RequestStats {
    pending: number;
    approved: number;
    rejected: number;
    cancelled: number;
}

// Service
export const academicRequestService = {
    // Student APIs

    /**
     * Get all available request types with deadline info
     */
    getRequestTypes: async (): Promise<AcademicRequestType[]> => {
        const response = await apiClient.get('/v1/academic-requests/types');
        return response.data;
    },

    /**
     * Check if a request can be submitted (deadline validation)
     */
    checkDeadline: async (requestType: string, classSectionId?: string): Promise<DeadlineCheckResponse> => {
        const params: Record<string, string> = { requestType };
        if (classSectionId) {
            params.classSectionId = classSectionId;
        }
        const response = await apiClient.get('/v1/academic-requests/check-deadline', { params });
        return response.data;
    },

    /**
     * Create a new academic request with optional file attachment
     */
    createRequest: async (payload: CreateAcademicRequestPayload, file?: File): Promise<AcademicRequest> => {
        const formData = new FormData();
        formData.append('request', new Blob([JSON.stringify(payload)], { type: 'application/json' }));
        if (file) {
            formData.append('file', file);
        }
        const response = await apiClient.post('/v1/academic-requests', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    /**
     * Get my requests (student)
     */
    getMyRequests: async (
        page = 0,
        size = 10,
        sort = 'createdAt,desc',
        filters?: {
            status?: string;
            requestType?: string;
        }
    ): Promise<PageResponse<AcademicRequest>> => {
        const params: Record<string, string | number> = { page, size, sort };
        if (filters?.status) params.status = filters.status;
        if (filters?.requestType) params.requestType = filters.requestType;

        const response = await apiClient.get('/v1/academic-requests/my-requests', {
            params
        });
        return response.data;
    },

    /**
     * Cancel my request (student)
     */
    cancelRequest: async (id: number): Promise<AcademicRequest> => {
        const response = await apiClient.put(`/v1/academic-requests/my-requests/${id}/cancel`);
        return response.data;
    },

    // Academic Staff APIs

    /**
     * Get all requests with filters (staff only)
     */
    getRequests: async (
        page = 0,
        size = 10,
        sort = 'createdAt,desc',
        filters?: {
            search?: string;
            status?: string;
            requestType?: string;
        }
    ): Promise<PageResponse<AcademicRequest>> => {
        const params: Record<string, string | number> = { page, size, sort };
        if (filters?.search) params.search = filters.search;
        if (filters?.status) params.status = filters.status;
        if (filters?.requestType) params.requestType = filters.requestType;

        const response = await apiClient.get('/v1/academic-requests', { params });
        return response.data;
    },

    /**
     * Get request statistics (staff only)
     */
    getStats: async (): Promise<RequestStats> => {
        const response = await apiClient.get('/v1/academic-requests/stats');
        return response.data;
    },

    /**
     * Get request by ID
     */
    getRequestById: async (id: number): Promise<AcademicRequest> => {
        const response = await apiClient.get(`/v1/academic-requests/${id}`);
        return response.data;
    },

    /**
     * Update request status (approve/reject) - staff only
     */
    updateStatus: async (id: number, status: 'APPROVED' | 'REJECTED', note?: string): Promise<AcademicRequest> => {
        const response = await apiClient.put(`/v1/academic-requests/${id}/status`, { status, note });
        return response.data;
    },
};

export default academicRequestService;

File: backend/src/main/java/com/fams/backend/service/AcademicRequestService.java
package com.fams.backend.service;

import com.fams.backend.dto.request.CreateAcademicRequestDTO;
import com.fams.backend.dto.response.AcademicRequestResponse;
import com.fams.backend.entity.AcademicRequest.AcademicRequestType;
import com.fams.backend.entity.AcademicRequest.RequestStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

public interface AcademicRequestService {

    /**
     * Create a new academic request
     * 
     * @param request   DTO containing request details
     * @param file      Optional file attachment
     * @param studentId ID of the student making the request
     * @return Created request response
     */
    AcademicRequestResponse createRequest(CreateAcademicRequestDTO request, MultipartFile file, Long studentId);

    /**
     * Get all requests for a student
     * 
     * @param studentId ID of the student
     * @param pageable  Pagination info
     * @return Page of requests
     */
    Page<AcademicRequestResponse> getRequestsByStudent(Long studentId, Pageable pageable);

    /**
     * Get all requests for a student with filters
     *
     * @param studentId   ID of the student
     * @param status      Filter by status
     * @param requestType Filter by type
     * @param pageable    Pagination info
     * @return Page of requests
     */
    Page<AcademicRequestResponse> getRequestsByStudent(Long studentId, RequestStatus status,
            AcademicRequestType requestType, Pageable pageable);

    /**
     * Get all requests with filters (for academic staff)
     * 
     * @param search      Search query
     * @param status      Filter by status
     * @param requestType Filter by type
     * @param pageable    Pagination info
     * @return Page of requests
     */
    Page<AcademicRequestResponse> getRequests(
            String search,
            RequestStatus status,
            AcademicRequestType requestType,
            Pageable pageable);

    /**
     * Get request by ID
     * 
     * @param id Request ID
     * @return Request response
     */
    AcademicRequestResponse getRequestById(Long id);

    /**
     * Update request status (approve/reject)
     * 
     * @param id         Request ID
     * @param status     New status
     * @param note       Approver note
     * @param approverId ID of the approver
     * @return Updated request response
     */
    AcademicRequestResponse updateRequestStatus(Long id, RequestStatus status, String note, Long approverId);

    /**
     * Cancel a pending request (by student)
     * 
     * @param id        Request ID
     * @param studentId ID of the student
     * @return Updated request response
     */
    AcademicRequestResponse cancelRequest(Long id, Long studentId);

    /**
     * Get request statistics
     * 
     * @return Map of status to count
     */
    Map<String, Long> getRequestStats();

    /**
     * Get all request types with deadline info
     * 
     * @param studentId ID of the student (to check class section for 3DA)
     * @return List of request type info
     */
    List<Map<String, Object>> getRequestTypes(Long studentId);

    /**
     * Check if request can be submitted (within deadline window)
     * 
     * @param requestType    Type of request
     * @param classSectionId Class section ID (for 3DA type)
     * @return Map with canSubmit flag and deadline info
     */
    Map<String, Object> checkDeadline(AcademicRequestType requestType, String classSectionId);
}
